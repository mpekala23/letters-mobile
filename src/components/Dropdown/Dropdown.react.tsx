import React, { createRef } from 'react';
import { Text, View, Animated, TouchableOpacity } from 'react-native';

import { Typography } from '@styles';
import Styles from './Dropdown.styles';

const DROPDOWN_HEIGHT = 100;
const ANIM_DURATION = 500;
const DROP_DURATION = 4000;

export enum DropType {
  Success = 'Success',
  Warning = 'Warning',
  Error = 'Error',
}

export interface DropNotif {
  type: DropType;
  message: string | JSX.Element;
  dynamic?: boolean;
  persist: boolean;
  duration: number;
  onPress?: () => void;
  id?: number;
}

interface State {
  dropped: boolean;
  height: Animated.Value;
  notifQ: DropNotif[];
  animating: boolean;
}

export class Dropdown extends React.Component<Record<string, unknown>, State> {
  private numNotifs = 1;

  constructor(props: Record<string, unknown>) {
    super(props);
    this.state = {
      dropped: false,
      height: new Animated.Value(-DROPDOWN_HEIGHT),
      notifQ: [],
      animating: false,
    };
    this.queueNotif = this.queueNotif.bind(this);
    this.flushNotif = this.flushNotif.bind(this);
    this.endNotif = this.endNotif.bind(this);
    this.renderNotif = this.renderNotif.bind(this);
  }

  queueNotif(notif: DropNotif): void {
    const newNotif = notif;
    newNotif.id = this.numNotifs;
    this.numNotifs += 1;
    this.setState(
      ({ notifQ }) => {
        const currentNotifs = [...notifQ];
        currentNotifs.push(notif);
        return {
          notifQ: currentNotifs,
        };
      },
      () => {
        this.flushNotif();
      }
    );
  }

  // convenient success dropdown
  queueSuccess({
    message,
    dynamic,
    onPress,
    persist,
    duration,
  }: {
    message: string | JSX.Element;
    dynamic?: boolean;
    onPress: () => void;
    persist: boolean;
    duration: number;
  }): void {
    const successNotif: DropNotif = {
      type: DropType.Success,
      message,
      dynamic,
      onPress,
      persist,
      duration,
    };
    this.queueNotif(successNotif);
  }

  // convenient success dropdown
  queueWarning({
    message,
    dynamic,
    onPress,
    persist,
    duration,
  }: {
    message: string | JSX.Element;
    dynamic?: boolean;
    onPress: () => void;
    persist: boolean;
    duration: number;
  }): void {
    const successNotif: DropNotif = {
      type: DropType.Warning,
      message,
      dynamic,
      onPress,
      persist,
      duration,
    };
    this.queueNotif(successNotif);
  }

  // convenient error dropdown
  queueError({
    message,
    dynamic,
    onPress,
    persist,
    duration,
  }: {
    message: string | JSX.Element;
    dynamic?: boolean;
    onPress: () => void;
    persist: boolean;
    duration: number;
  }): void {
    const errorNotif: DropNotif = {
      type: DropType.Error,
      message,
      dynamic,
      onPress,
      persist,
      duration,
    };
    this.queueNotif(errorNotif);
  }

  flushNotif(): void {
    if (
      this.state.dropped ||
      this.state.animating ||
      this.state.notifQ.length === 0
    )
      return;
    this.setState({ dropped: true, animating: true }, () => {
      const currentId = this.state.notifQ[0].id || -1;
      Animated.timing(this.state.height, {
        toValue: 0,
        duration: ANIM_DURATION,
        useNativeDriver: false,
      }).start(() => {
        // when the animation finishes
        this.setState({ animating: false });
        if (!this.state.notifQ[0].persist)
          setTimeout(() => {
            this.endNotif(currentId);
          }, this.state.notifQ[0].duration);
      });
    });
  }

  endNotif(id: number): void {
    if (
      !this.state.dropped ||
      this.state.animating ||
      this.state.notifQ.length === 0 ||
      this.state.notifQ[0].id !== id
    )
      return;
    this.setState({ animating: true }, () => {
      Animated.timing(this.state.height, {
        toValue: -DROPDOWN_HEIGHT,
        duration: ANIM_DURATION,
        useNativeDriver: false,
      }).start(() => {
        // when the animation finishes
        this.setState(
          (prevState) => {
            const newState = { ...prevState };
            newState.notifQ.splice(0, 1);
            newState.dropped = false;
            newState.animating = false;
            return newState;
          },
          () => {
            // when the notifQ has been updated
            if (this.state.notifQ.length > 0) {
              this.flushNotif();
            }
          }
        );
      });
    });
  }

  renderNotif(): JSX.Element {
    const notif = this.state.notifQ[0];
    if (!notif) return <View />;
    let typeBackground = Styles.errorBackground;
    let typeText = Styles.errorText;
    if (notif.type === DropType.Warning) {
      typeBackground = Styles.warningBackground;
      typeText = Styles.warningText;
    } else if (notif.type === DropType.Success) {
      typeBackground = Styles.successBackground;
      typeText = Styles.successText;
    }
    return this.state.dropped ? (
      <TouchableOpacity
        testID="touchable"
        style={[Styles.commonBackground, typeBackground]}
        onPress={() => {
          if (notif.onPress) notif.onPress();
          if (notif.id) this.endNotif(notif.id);
        }}
        activeOpacity={0.9}
      >
        {notif.dynamic ? (
          notif.message
        ) : (
          <Text style={[Typography.FONT_MEDIUM, typeText]}>
            {notif.message}
          </Text>
        )}
      </TouchableOpacity>
    ) : (
      <View />
    );
  }

  render(): JSX.Element {
    return (
      <Animated.View
        style={{
          position: 'absolute',
          zIndex: 999,
          elevation: 999,
          top: this.state.height,
          width: '100%',
        }}
      >
        {this.renderNotif()}
      </Animated.View>
    );
  }
}

const dropdownRef = createRef<Dropdown>();
const DropdownInstance = (): JSX.Element => (
  <Dropdown ref={dropdownRef} key="Dropdown" />
);

export function dropdownSuccess({
  message,
  dynamic,
  onPress,
  persist,
  duration,
}: {
  message?: string | JSX.Element;
  dynamic?: boolean;
  onPress?: () => void;
  persist?: boolean;
  duration?: number;
}): void {
  if (dropdownRef.current)
    dropdownRef.current.queueSuccess({
      message: message || '',
      dynamic,
      onPress: onPress || (() => null),
      persist: persist || false,
      duration: duration || DROP_DURATION,
    });
}

export function dropdownWarning({
  message,
  dynamic,
  onPress,
  persist,
  duration,
}: {
  message?: string | JSX.Element;
  dynamic?: boolean;
  onPress?: () => void;
  persist?: boolean;
  duration?: number;
}): void {
  if (dropdownRef.current)
    dropdownRef.current.queueWarning({
      message: message || '',
      dynamic,
      onPress: onPress || (() => null),
      persist: persist || false,
      duration: duration || DROP_DURATION,
    });
}

export function dropdownError({
  message,
  dynamic,
  onPress,
  persist,
  duration,
}: {
  message?: string | JSX.Element;
  dynamic?: boolean;
  onPress?: () => void;
  persist?: boolean;
  duration?: number;
}): void {
  if (dropdownRef.current)
    dropdownRef.current.queueError({
      message: message || '',
      dynamic,
      onPress: onPress || (() => null),
      persist: persist || false,
      duration: duration || DROP_DURATION,
    });
}

export default DropdownInstance;
