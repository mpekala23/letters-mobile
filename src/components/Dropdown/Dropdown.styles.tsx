import { StyleSheet } from 'react-native';
import { STATUS_BAR_HEIGHT } from '@utils';
import { Colors } from '@styles';

export default StyleSheet.create({
  commonBackground: {
    flex: 1,
    zIndex: 999,
    elevation: 999,
    marginTop: STATUS_BAR_HEIGHT,
    marginHorizontal: 5,
    padding: STATUS_BAR_HEIGHT,
    justifyContent: 'center',
    borderRadius: 12,
  },
  successBackground: {
    backgroundColor: Colors.DROPDOWN_SUCCESS_BACKGROUND,
  },
  successText: {
    fontSize: 16,
    color: Colors.DROPDOWN_SUCCESS_TEXT,
  },
  warningBackground: {
    backgroundColor: Colors.DROPDOWN_WARNING_BACKGROUND,
  },
  warningText: {
    fontSize: 16,
    color: Colors.DROPDOWN_WARNING_TEXT,
  },
  errorBackground: {
    backgroundColor: Colors.DROPDOWN_ERROR_BACKGROUND,
  },
  errorText: {
    fontSize: 16,
    color: Colors.DROPDOWN_ERROR_TEXT,
  },
});
