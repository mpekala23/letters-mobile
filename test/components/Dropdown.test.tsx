import React from 'react';
import Dropdown, {
  dropdownInfo,
  dropdownSuccess,
  dropdownWarning,
  dropdownError,
} from '@components/Dropdown/Dropdown.react';
import { Colors } from '@styles';
import { render, toJSON } from '@testing-library/react-native';

jest.useFakeTimers();

const setup = () => {
  return {
    ...render([Dropdown()]),
  };
};

describe('Dropdown component', () => {
  it('should match snapshot', () => {
    const { container } = setup();
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it('should implement info dropdowns', () => {
    const { getByText, getByTestId } = setup();
    dropdownInfo('title', 'body');
    const bodyText = getByText('body');
    expect(bodyText).toBeDefined();
    const touchable = getByTestId('touchable');
    expect(touchable.props.style[1].backgroundColor).toBe(Colors.AMEELIO_BLUE);
  });

  it('should implement success dropdowns', () => {
    const { getByText, getByTestId } = setup();
    dropdownSuccess('title', 'body');
    const bodyText = getByText('body');
    expect(bodyText).toBeDefined();
    const touchable = getByTestId('touchable');
    expect(touchable.props.style[1].backgroundColor).toBe(Colors.SUCCESS);
  });

  it('should implement warning dropdowns', () => {
    const { getByText, getByTestId } = setup();
    dropdownWarning('title', 'body');
    const bodyText = getByText('body');
    expect(bodyText).toBeDefined();
    const touchable = getByTestId('touchable');
    expect(touchable.props.style[1].backgroundColor).toBe(Colors.WARNING);
  });

  it('should implement error dropdowns', () => {
    const { getByText, getByTestId } = setup();
    dropdownError('title', 'body');
    const bodyText = getByText('body');
    expect(bodyText).toBeDefined();
    const touchable = getByTestId('touchable');
    expect(touchable.props.style[1].backgroundColor).toBe(Colors.ERROR);
  });

  it('should queue dropdowns', () => {
    const { getByText, queryByText } = setup();
    dropdownError('title1', 'body1');
    dropdownError('title2', 'body2');
    expect(getByText('body1')).toBeDefined();
    expect(queryByText('body2')).toBe(null);
  });
});
