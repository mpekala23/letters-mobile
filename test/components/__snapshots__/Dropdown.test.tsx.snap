// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "elevation": 999,
        "position": "absolute",
        "top": -100,
        "width": "100%",
        "zIndex": 999,
      }
    }
  >
    <View />
  </View>
</View>
`;
